### PR TITLE
common: periodically dump the priority queue to a JSON file

### PR DESCRIPTION
**Pull request by FrostyX**
Many times in the past I wanted to see how exactly the priority queue looks like at a given time but we didn't have a way of doing it. I wanted to implement some script for displaying it, but that's not possible because we AFAIK keep all the priority queue state in a variable (which of course is not accessible from the outside). So what about periodically dumping the queue into a JSON file by the `Dispatcher` itself?

I think this would be good enough for all the times that I needed it in the past but in case it was not sufficient, this would allow us to implement any kind of script to visualize the queue any way we want.

### CODE REVIEW

**Comment by FrostyX** [comment]
If you dislike the idea in general, we can drop the PR. No hard feelings :D

File: common/copr_common/dispatcher.py [hunk 1]
```diff
@@ -126,3 +129,20 @@ def run(self):
 
         # reset the title
         self._update_process_title()
+
+    def dump_queue(self, worker_manager):
+        """
+        Dump the current priority queue state into a JSON file
```
Comment (xsuchy) [inline common/copr_common/dispatcher.py:135]: Can you put in the docstring the reason why we are doing this? So your future yourself do not worry it is dead code.

Comment (praiskup) [reply common/copr_common/dispatcher.py:135]: It is not a dead code actually ... it is dumped in every single iteration, which raises my worries.  Could we have an opt-in knob, and document there when we need it?

File: common/copr_common/dispatcher.py [hunk 2]
```diff
@@ -126,3 +129,20 @@ def run(self):
...
+        queue = []
+        for entry in worker_manager.tasks.prio_queue:
+            # This should never happen but let's make sure we don't traceback
+            if not isinstance(entry, list) or len(entry) != 3:
+                return
+
+            priority, count, task = entry
+            queue.append([priority, count, task.id])
+
+        path = "/tmp/copr-{0}-queue.json".format(self.name)
```
Comment (praiskup) [inline common/copr_common/dispatcher.py:146]: Not sure if `/tmp` is the ideal place, perhaps /var/log?

**Comment by FrostyX** [comment]
Updated, PTAL

File: common/copr_common/dispatcher.py [hunk 3]
```diff
@@ -126,3 +130,29 @@ def run(self):
...
         self._update_process_title()
+
+    def dump_queue(self, worker_manager):
+        """
+        Dump the current priority queue state into a JSON file.
+        This is useful when we want to see the exact state of the queue from
+        backend's perspective (e.g. when some tasks are not being processed).
+        """
+        # This function is usefull for debugging, there is no need to run it
+        # all the time
```
Comment (praiskup) [inline common/copr_common/dispatcher.py:141]: s/usefull/useful

**Review by praiskup — APPROVED** [review]

**Pull request merged**